LinkedIdeas/CanvasView.swift

```diff
@@ -42,23 +47,126 @@ class CanvasView: NSView {
     NSColor.whiteColor().set()
     NSBezierPath(rect: bounds).fill()
     
-    for concept in concepts {
-      if !concept.added {
-        sprint("add concept view")
-        let conceptView = ConceptView(frame: concept.rect)
-        conceptView.concept = concept
-        concept.added = true
-        addSubview(conceptView)
-        conceptView.canvas = self
-      }
+    if mode == Mode.Links {
+      drawCreationLinkArrow()
+      for link in links { addLinkView(link) }
+    }
+    
+    if mode == Mode.Concepts {
+      for concept in concepts { addConceptView(concept) }
     }
+    
+    // for point in clicks { drawCenteredDotAtPoint(point, color: NSColor.cyanColor()) }
   }
   
+  // MARK: - Drawing Functions
+  
+  func drawCreationLinkArrow() {
+    if let arrowStart = arrowStart, arrowEnd = arrowEnd {
+      sprint("render arrow")
+      NSColor.blackColor().set()
+      let path = NSBezierPath()
+      path.moveToPoint(arrowStart)
+      path.lineToPoint(arrowEnd)
+      path.stroke()
+    }
+  }
+  
+  // MARK: - Mouse Events
+  
   override func mouseDown(theEvent: NSEvent) {
     sprint("canvasView: mouse down")
-    delegate?.singleClick(theEvent)
+    let point = convertPoint(theEvent.locationInWindow, fromView: nil)
+    clicks.append(point)
+    
+    for concept in concepts {
+      concept.editing = false
+    }
+    
+    if mode == Mode.Concepts {
+      delegate?.singleClick(theEvent)
+    } else {
+      arrowStart = convertPoint(theEvent.locationInWindow, fromView: nil)
+    }
   }
   
+  func mouseDownFromConcept(theEvent: NSEvent) {
+    if mode == Mode.Links {
+      arrowStart = convertPoint(theEvent.locationInWindow, fromView: nil)
+    }
+  }
+  
+  override func mouseDragged(theEvent: NSEvent) {
+    if mode == Mode.Links {
+      sprint("mouse dragged")
+      arrowEnd = convertPoint(theEvent.locationInWindow, fromView: nil)
+    }
+  }
+  
+  override func mouseUp(theEvent: NSEvent) {
+    if mode == Mode.Links {
+      sprint("mouse up")
+      if let targetConceptIdentifier = targetConceptIdentifier, originConceptIdentifier = originConceptIdentifier {
```

Comment:
If you remove the arrows anyways, you can put `removeArrow()` outside and concatenate the two ifs, like:

```swift
if let targetConceptIdentifier = targetConceptIdentifier, originConceptIdentifier = originConceptIdentifier where  originConceptIdentifier != targetConceptIdentifier {  }
```